### PR TITLE
AnalyticsService with events and actions

### PR DESCRIPTION
**Pull request by kbats183**

### CODE REVIEW

File: backend/src/main/kotlin/org/icpclive/widget/PresetsController.kt [hunk 1]
```diff
@@ -69,7 +69,7 @@ class PresetsController<SettingsType : ObjectSettings, OverlayWidgetType : TypeW
     }
 
     suspend fun delete(id: Int) = mutex.withLock {
-        val preset = findById(id)
+        val preset = findByIdOrNull(id) ?: return@withLock
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/widget/PresetsController.kt:72]: It's ok to just return here

File: backend/src/main/kotlin/org/icpclive/service/QueueService.kt [hunk 2]
```diff
@@ -69,21 +82,40 @@ class QueueService {
...
                     val run = event.run
-                    val currentTime = contestInfoFlow.value.currentContestTime.takeIf { it != firstEventTime } ?: run.time
+                    val currentTime =
+                        contestInfoFlow.value.currentContestTime.takeIf { it != firstEventTime } ?: run.time
                     logger.debug("Receive run $run")
                     lastUpdateTime[run.id] = currentTime
                     if (run.id in runs || contestInfoFlow.value.currentContestTime <= currentTime + run.timeInQueue) {
-                        resultFlow.emit(if (run.id in runs) ModifyRunInQueueEvent(run) else AddRunToQueueEvent(run))
-                        runs[run.id] = run
+                        modifyRun(run)
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/service/QueueService.kt:90]: I think we need to preserve isFeaturedRun here. As now update of featured run will make it unfeatured

File: backend/src/main/kotlin/org/icpclive/service/QueueService.kt [hunk 3]
```diff
@@ -43,14 +46,23 @@ class QueueService {
...
         runs.remove(run.id)
+        removedRuns[run.id] = run
         resultFlow.emit(RemoveRunFromQueueEvent(run))
     }
 
     private val RunInfo.timeInQueue
-        get() = if (isFirstSolvedRun) FIRST_TO_SOLVE_WAIT_TIME else WAIT_TIME
-
+        get() = when {
+            isFeaturedRun -> FEATURED_WAIT_TIME
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/service/QueueService.kt:62]: And what if featured run is also first to solve?

File: backend/src/main/kotlin/org/icpclive/data/DataBus.kt [hunk 4]
```diff
@@ -16,11 +17,16 @@ object DataBus {
     val contestInfoFlow = CompletableDeferred<Flow<ContestInfo>>()
     val mainScreenFlow = CompletableDeferred<Flow<MainScreenEvent>>()
     val queueFlow = CompletableDeferred<Flow<QueueEvent>>()
+    val queueFeaturedRunsFlow = MutableSharedFlow<Int>( // flow of run ids that need to be braking news
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/data/DataBus.kt:20]: Can it be registered by service filling it? Or you think it should be filled by several services?

File: backend/src/main/kotlin/org/icpclive/api/Objects.kt [hunk 5]
```diff
@@ -163,14 +164,20 @@ data class ProblemSolutionsStatistic(val success: Int, val wrong: Int, val pendi
 data class SolutionsStatistic(val stats: List<ProblemSolutionsStatistic>)
 
 @Serializable
-sealed class AnalyticsEvent {
+class AnalyticsCompanionPreset(
+    val presetId: Int,
+    val expirationTime: Instant?,
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/api/Objects.kt:169]: Should have @Serializable(with=), and better to have SerialName with units.

File: backend/src/main/kotlin/org/icpclive/service/QueueService.kt [hunk 6]
```diff
@@ -69,21 +82,40 @@ class QueueService {
...
+                    if (run == null) {
+                        logger.warn("There is no run with id ${event.runId} for make it featured")
+                        logger.info(runs.values.toString())
+                        return@collect
+                    }
+                    modifyRun(run.copy(isFeaturedRun = true))
+                    lastUpdateTime[run.id] =
+                        contestInfoFlow.value.currentContestTime.takeIf { it != firstEventTime } ?: run.time
+                }
+                is MakeNotFeatured -> {
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/service/QueueService.kt:104]: I failed to find anybody who triggers this event. Is it ok?

**Review by kunyavskiy — CHANGES_REQUESTED** [review]

File: backend/src/main/kotlin/org/icpclive/data/DataBus.kt (hunk 4 above)
Comment (kbats183) [reply backend/src/main/kotlin/org/icpclive/data/DataBus.kt:20]: I think Analytics Service is optional and we will disable it sometimes. Therefore, Queue may be no depend on analytics

File: backend/src/main/kotlin/org/icpclive/service/QueueService.kt (hunk 3 above)
Comment (kbats183) [reply backend/src/main/kotlin/org/icpclive/service/QueueService.kt:62]: have i to sum times?

Comment (kunyavskiy) [reply backend/src/main/kotlin/org/icpclive/service/QueueService.kt:62]: No idea, but current behaviour is strange. Many
Maybe @lperovskaya knows how it should work.

File: backend/src/main/kotlin/org/icpclive/data/DataBus.kt (hunk 4 above)
Comment (kunyavskiy) [reply backend/src/main/kotlin/org/icpclive/data/DataBus.kt:20]: Okey, let's keep it for now, I'll think about it later.

File: backend/src/main/kotlin/org/icpclive/service/QueueService.kt [hunk 7]
```diff
@@ -107,7 +112,8 @@ class QueueService {
                         logger.warn("There is no run with id ${event.runId} for make it not featured")
                         return@collect
                     }
-                    modifyRun(run.copy(isFeaturedRun = false))
+                    featuredRuns += run.id
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/service/QueueService.kt:115]: Probably, you wanted to remove, not add here

File: backend/src/main/kotlin/org/icpclive/api/Objects.kt [hunk 8]
```diff
@@ -163,14 +164,26 @@ data class ProblemSolutionsStatistic(val success: Int, val wrong: Int, val pendi
 data class SolutionsStatistic(val stats: List<ProblemSolutionsStatistic>)
 
 @Serializable
-sealed class AnalyticsEvent {
+data class AnalyticsCompanionPreset(
+    val presetId: Int,
+    @SerialName("expirationTimeMs")
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/api/Objects.kt:169]: Let it be expirationTimeUnixMs for being more consistent.

File: backend/src/main/kotlin/org/icpclive/admin/Analytics.kt [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+    fun Route.presetWidget(
+        name: String,
+        showAction: (id: String, ttlMs: Long?) -> AnalyticsAction,
+        hideAction: (id: String) -> AnalyticsAction
+    ) {
+        route("/$name") {
+            post {
+                call.adminApiAction {
+                    actionsFlow.emit(showAction(call.id(), call.request.queryParameters["ttl"]?.toLong()))
```
Comment (kunyavskiy) [inline backend/src/main/kotlin/org/icpclive/admin/Analytics.kt:27]: Let's parse TTL to kotlin.time.Duration here and work with it in the main code?

**Review by kunyavskiy — APPROVED** [review]

**Pull request merged**